From 7b0d50e4b1d16fafd55876260701e04279a7489d Mon Sep 17 00:00:00 2001
From: Sebastian Pipping <sebastian@pipping.org>
Date: Mon, 26 Oct 2026 17:39:32 +0100
Subject: [PATCH] lib: Fix initial non-final empty subparser call with a
 handler

[Hades: from libexpat pull request #1444, commit 7b0d50e4b1d16fafd55876260701e04279a7489d,
 not merged as of 2026-10-28; taken for CVE-2026-117610 and CVE-2026-118370.]

---
 lib/xmlparse.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/xmlparse.c b/lib/xmlparse.c
index a04ffde..d2edf9a 100644
--- a/lib/xmlparse.c
+++ b/lib/xmlparse.c
@@ -3537,6 +3537,11 @@ externalEntityInitProcessor2(XML_Parser parser, const char *start,
     }
     parser->m_eventPtr = start;
     return XML_ERROR_PARTIAL_CHAR;
+  case XML_TOK_NONE:
+    if (! parser->m_parsingStatus.finalBuffer) {
+      *endPtr = start;
+    }
+    return XML_ERROR_NONE;
   }
   parser->m_processor = externalEntityInitProcessor3;
   return externalEntityInitProcessor3(parser, start, end, endPtr);
-- 
2.53.0

